src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -0,0 +1,657 @@
+
+                    void finish(Downstream<? super RR> c) {
+                        if (leftFinisher != Gatherer.<A, R>defaultFinisher())
+                            leftFinisher.accept(leftState, r -> rightIntegrate(r, c));
+                        if (rightFinisher != Gatherer.<AA, RR>defaultFinisher())
+                            rightFinisher.accept(rightState, c);
+                    }
+
+                    /*
+                     * Currently we use the following to ferry elements from
+                     * the left Gatherer to the right Gatherer, but we create
+                     * the Gatherer.Downstream as a lambda which means that
+                     * the default implementation of `isKnownDone()` is used.
+                     *
+                     * If it is determined that we want to be able to support
+                     * the full interface of Gatherer.Downstream then we have
+                     *  the following options:
+                     *    1. Have State implement Downstream<? super R>
+                     *       and store the passed in Downstream<? super RR>
+                     *       downstream as an instance field in integrate()
+                     *       and read it in push(R r).
+                     *    2. Allocate a new Gatherer.Downstream<? super R> for
+                     *       each invocation of integrate() which might prove
+                     *       costly.
+                     */
+                    public boolean rightIntegrate(R r, Downstream<? super RR> downstream) {
+                        // The following logic is highly performance sensitive
+                        return (rightGreedy || rightProceed)
+                                && (rightIntegrator.integrate(rightState, r, downstream)
+                                || rightGreedy
+                                || (rightProceed = false));
+                    }
+                }
+
+                return new GathererImpl<T, State, RR>(
+                        State::new,
+                        (leftGreedy && rightGreedy)
+                                ? Integrator.<State, T, RR>ofGreedy(State::integrate)
+                                : Integrator.<State, T, RR>of(State::integrate),
+                        (leftCombiner == Gatherer.defaultCombiner()
+                        || rightCombiner == Gatherer.defaultCombiner())
+                                ? Gatherer.defaultCombiner()
+                                : State::joinLeft,
+                        (leftFinisher == Gatherer.<A, R>defaultFinisher()
+                        && rightFinisher == Gatherer.<AA, RR>defaultFinisher())
+                                ? Gatherer.defaultFinisher()
+                                : State::finish
+                );
+            }
+        }
+    }
+
+    // Public built-in Gatherers and factory methods for them
+
+    /**
+     * Gathers elements into fixed-size windows. The last window may contain
```

Comment:
Returns a ...

Reply:
You mean "Returns a Gatherer which gathers elements into ..." ?

Reply:
Is "window" defined anywhere? If you are doing another pass over the docs then the windowXXX methods could be expanded a bit to explain this grouping concept. Also it could say that the last windows (if it exists, the empty stream is the corner case) has at least one element.

Reply:
@AlanBateman Good point. I may have assumed too much. Perhaps something along the lines of:

"Returns a Gatherer which gathers elements into windows -- encounter-ordered groups of elements -- …" ?